### PR TITLE
Add e2e tests for diagrams and embeddings

### LINKED ISSUE #1022: Add E2E tests for "special" embeddings
**Which part of the project should be enhanced?**
The different special embeddings (like YouTube, pdf-to-link, Gist, ...) should get E2E tests to ensure that they stay intact.

**Is your enhancement request related to a problem? Please describe.**
The pdf-to-link replacer seems to not work anymore after some dependency updates.

### PR DESCRIPTION
**Pull request by ErikMichelson**
### Component/Part
E2E tests

### Description
This PR adds e2e tests for diagrams and embeddings. As most of the diagrams only yield a `svg` element into the DOM, the tests can ensure that some svg element is visible, but can't ensure that it really belongs to the tested diagram.

### Steps

- [x] Added / updated tests
- [x] I read the [contribution documentation](https://github.com/hedgedoc/react-client/blob/main/CONTRIBUTING.md) and signed-off my commits to accept the DCO.

### Related Issue(s)
Closes #1022

### CODE REVIEW

**Review by DerMolly — APPROVED** [review]

File: src/components/markdown-renderer/replace-components/youtube/youtube-frame.tsx
Comment (mrdrogdrog) [inline src/components/markdown-renderer/replace-components/youtube/youtube-frame.tsx:18]: Please correct the image url as well.

File: cypress/support/config.ts
Comment (mrdrogdrog) [inline cypress/support/config.ts:48]: Please don't use a live plantuml server for the test. It slows down the test runner, requires access to the internet and the test could fail if plantuml is down. How about a `cy.intercept` mock?

File: cypress/integration/shortcodes.spec.ts
Comment (mrdrogdrog) [inline cypress/integration/shortcodes.spec.ts:20]: what about the youtube shortcode?

File: cypress/integration/diagrams.spec.ts
Comment (mrdrogdrog) [inline cypress/integration/diagrams.spec.ts:12]: "Renders diagram markmap" is not a valid sentence. The describe+it calls should build a valid sentence like "Short code for pdf renders a plain link". That's why it's called "it" e.g. "It renders a plain link".
What about
"Diagram embeddings" + "for markdown renders a svg".

File: cypress/integration/diagrams.spec.ts
Comment (mrdrogdrog) [inline cypress/integration/diagrams.spec.ts:22]: Is the div or the canvas relevant here? The test shouldn't care about how vega-lite renders the diagram.

File: cypress/integration/diagrams.spec.ts
Comment (mrdrogdrog) [inline cypress/integration/diagrams.spec.ts:36]: div

File: cypress/integration/diagrams.spec.ts
Comment (mrdrogdrog) [inline cypress/integration/diagrams.spec.ts:50]: div

File: cypress/integration/diagrams.spec.ts
Comment (mrdrogdrog) [inline cypress/integration/diagrams.spec.ts:57]: is the `table` selector necessary?

File: cypress/integration/diagrams.spec.ts
Comment (mrdrogdrog) [inline cypress/integration/diagrams.spec.ts:65]: Don't rely on titles. I think the URL is a more secure indicator. (Combined with the annotation below)

File: cypress/integration/linkEmbedder.spec.ts
Comment (mrdrogdrog) [inline cypress/integration/linkEmbedder.spec.ts:19]: Why do you rely on the title and not on the explicit URL you set above?

File: cypress/integration/linkEmbedder.spec.ts
Comment (mrdrogdrog) [inline cypress/integration/linkEmbedder.spec.ts:15]: And what if we switch from span to div? That would invalidate this test. The `span` selector is redundant here.

File: cypress/integration/linkEmbedder.spec.ts
Comment (mrdrogdrog) [inline cypress/integration/linkEmbedder.spec.ts:25]: see above.

**Review by mrdrogdrog — CHANGES_REQUESTED** [review]

File: cypress/integration/diagrams.spec.ts
Comment (mrdrogdrog) [inline cypress/integration/diagrams.spec.ts:15]: Tbh i'm not happy with just searching for any svg. If I switch markmap with e.g. graphviz, then this test would still be happy. Please use an (additional) more specific selector. But don't add a new CSS class. Use `data-cy` like we do in other tests.

File: cypress/integration/diagrams.spec.ts
Comment (mrdrogdrog) [inline cypress/integration/diagrams.spec.ts:29]: see above

File: cypress/integration/diagrams.spec.ts
Comment (mrdrogdrog) [inline cypress/integration/diagrams.spec.ts:43]: see above

File: cypress/integration/linkEmbedder.spec.ts
Comment (mrdrogdrog) [inline cypress/integration/linkEmbedder.spec.ts:25]: Can you add a test for the one-click-embedding?

**Review by mrdrogdrog — CHANGES_REQUESTED** [review]

File: cypress/integration/diagrams.spec.ts
Comment (ErikMichelson) [reply cypress/integration/diagrams.spec.ts:12]: Well, "Short code for pdfs" isn't a sentence as well, just to name one example.
A complete sentence seems a bit verbose. What about the following schema?

Diagram embedding renders correctly: <markmap|vega-lite|...>

File: cypress/integration/diagrams.spec.ts
Comment (ErikMichelson) [reply cypress/integration/diagrams.spec.ts:22]: The div alone does not say that the diagram is rendered. Matching a canvas alone is a bit too vague.

File: cypress/integration/diagrams.spec.ts
Comment (ErikMichelson) [reply cypress/integration/diagrams.spec.ts:36]: Same as above

File: cypress/integration/diagrams.spec.ts
Comment (ErikMichelson) [reply cypress/integration/diagrams.spec.ts:50]: Same as above

File: cypress/integration/diagrams.spec.ts
Comment (ErikMichelson) [reply cypress/integration/diagrams.spec.ts:57]: Maybe not, but isn't it better to fence the test-element stricter than lesser?

File: cypress/integration/linkEmbedder.spec.ts
Comment (ErikMichelson) [reply cypress/integration/linkEmbedder.spec.ts:25]: What do you mean with this?

File: cypress/integration/linkEmbedder.spec.ts
Comment (ErikMichelson) [reply cypress/integration/linkEmbedder.spec.ts:19]: Because the URL is a dynamically generated base64 url of our iframe html code.

File: cypress/integration/diagrams.spec.ts
Comment (DerMolly) [reply cypress/integration/diagrams.spec.ts:36]: I think Tilman is suggesting to use `.mermaid > svg` or just `.mermaid` for the test. I guess if one wants the svg depends on how the error case looks. If in case of an error the output would also contain some HTML Tag with the class `.mermaid`, but not an `svg` this should be included…

File: cypress/integration/diagrams.spec.ts
Comment (mrdrogdrog) [reply cypress/integration/diagrams.spec.ts:22]: Just use the class?

File: cypress/integration/diagrams.spec.ts
Comment (mrdrogdrog) [reply cypress/integration/diagrams.spec.ts:36]: Then you should just use the class or use a data-cy tag. We're testing the replacer, not mermaid.

File: cypress/integration/diagrams.spec.ts
Comment (mrdrogdrog) [reply cypress/integration/diagrams.spec.ts:50]: same as above

File: cypress/integration/linkEmbedder.spec.ts
Comment (mrdrogdrog) [reply cypress/integration/linkEmbedder.spec.ts:19]: Well. fair. But then I would rather use a data-cy attribute. The title can change at any time.

File: cypress/integration/linkEmbedder.spec.ts
Comment (mrdrogdrog) [reply cypress/integration/linkEmbedder.spec.ts]: That the one-click-embedding component reacts to clicks, renders a preview image, loads the content behind it, etc.

File: src/components/markdown-renderer/replace-components/youtube/youtube-frame.tsx
Comment (ErikMichelson) [reply src/components/markdown-renderer/replace-components/youtube/youtube-frame.tsx]: Done

File: cypress/integration/shortcodes.spec.ts
Comment (ErikMichelson) [reply cypress/integration/shortcodes.spec.ts:20]: Added

File: cypress/support/config.ts
Comment (ErikMichelson) [reply cypress/support/config.ts:48]: How do you want to mock PlantUML? Returning a predefined "diagram" image url?

File: cypress/integration/linkEmbedder.spec.ts
Comment (ErikMichelson) [reply cypress/integration/linkEmbedder.spec.ts]: Done

File: cypress/integration/linkEmbedder.spec.ts
Comment (ErikMichelson) [reply cypress/integration/linkEmbedder.spec.ts]: Done

File: cypress/integration/diagrams.spec.ts
Comment (ErikMichelson) [reply cypress/integration/diagrams.spec.ts]: `data-cy` added

File: cypress/integration/diagrams.spec.ts
Comment (ErikMichelson) [reply cypress/integration/diagrams.spec.ts]: `data-cy` added

File: cypress/integration/diagrams.spec.ts
Comment (ErikMichelson) [reply cypress/integration/diagrams.spec.ts]: `data-cy` added

File: cypress/integration/linkEmbedder.spec.ts
Comment (ErikMichelson) [reply cypress/integration/linkEmbedder.spec.ts]: If you look a few lines below your annotation, you'll find the click-reaction and content-loading. Just the preview image needs to be tested yet.

File: cypress/support/config.ts
Comment (mrdrogdrog) [reply cypress/support/config.ts:48]: Yes

Comment (ErikMichelson) [reply cypress/support/config.ts]: Done

File: cypress/integration/diagrams.spec.ts
Comment (ErikMichelson) [reply cypress/integration/diagrams.spec.ts]: Done

File: cypress/integration/diagrams.spec.ts
Comment (ErikMichelson) [reply cypress/integration/diagrams.spec.ts]: Changed

File: cypress/integration/linkEmbedder.spec.ts
Comment (ErikMichelson) [reply cypress/integration/linkEmbedder.spec.ts]: Uses data-cy attribute now

File: cypress/integration/linkEmbedder.spec.ts
Comment (ErikMichelson) [reply cypress/integration/linkEmbedder.spec.ts]: Tests for preview images added

File: cypress/integration/diagrams.spec.ts
Comment (mrdrogdrog) [inline cypress/integration/diagrams.spec.ts]: I'm still not sure if it's a good idea to expect elements, that we don't control or create. The SVGs and the canvas you're looking for here are not part of our code, but of the 3rd-party rendering libs themself.
I understand, that you want to test if our code triggers the lib code but imho this is useless as long as we can't verify what the user is seeing. Also if e.g. abcjs switches from svg to canvas, then this test will fail and we still don't know if the output is correct.

File: cypress/integration/linkEmbedder.spec.ts
Comment (mrdrogdrog) [inline cypress/integration/linkEmbedder.spec.ts:27]: I think we should find a way to test the click-shield-component itself instead of using a specific implementation. But I don't know how. Maybe there are unit tests for React? :thinking:  Can you add a todo here for this?

**Review by mrdrogdrog — CHANGES_REQUESTED** [review]

File: cypress/integration/diagrams.spec.ts
Comment (mrdrogdrog) [reply cypress/integration/diagrams.spec.ts]: Well that's true. Okay, how about this: Describe blocks don't have to be part of the senteces, but should specify the parent describe block. But I insist, that the it-calls build correct senteces. It is a convention in testing.

File: cypress/integration/diagrams.spec.ts
Comment (ErikMichelson) [reply cypress/integration/diagrams.spec.ts]: It's true that we don't control the specific elements. But we sometimes control the parent container that has the class name or data-cy attribute and show for example error messages inside it on failures.
When just testing for the parent container, we don't make sure that the library really runs. That's not end-to-end testing but end-to-one-step-before-end.
While it's also true that maybe one day a dependency changes their rendering method, it's actually good that the test fails in that case. It reminds us to check whether the rendering works still after the dependency restructuring. And changing one selector in a cypress test inside a renovate PR isn't a major deal.

File: cypress/integration/linkEmbedder.spec.ts
Comment (ErikMichelson) [reply cypress/integration/linkEmbedder.spec.ts:27]: Sounds almost like [storybook tests](https://storybook.js.org/docs/react/workflows/testing-with-storybook).

I added a TODO entry.

File: cypress/integration/diagrams.spec.ts
Comment (ErikMichelson) [reply cypress/integration/diagrams.spec.ts]: It's important to have a good structure.
And a complete sentence might read good in single occasions and when you read it alone standing. But if you have dozens of log lines - like "Component xabc renders as expected", "Component ydfghj renders as expected" - it's much harder to get an overview of the actual component tested, especially as similar words are not aligned beneath each other.

I therefore stay at the position that test labels like "renders as expected: xabc", "renders as expected: ydfghj" fulfill the same purpose while being easier to read.

Comment (mrdrogdrog) [reply cypress/integration/diagrams.spec.ts]: I'm seeing this the opposite way. For me a senteces if much better readable. Especially in the output.

Sry but then I can't approve this, if we don't agree on a syntax, because then some tests are names like you want and some like I want and maybe @DerMolly also has another opinon.

File: cypress/integration/diagrams.spec.ts
Comment (mrdrogdrog) [reply cypress/integration/diagrams.spec.ts]: Tbh actually also doing an "end-to-one-step-before-end" test, because you're not checking if the lib renders the output correctly. To be a complete end to end test, you have to check if e.g. some words are contained in the output.

Comment (mrdrogdrog) [reply cypress/integration/diagrams.spec.ts]: We don't know if the svg you're checking contains any reasonable content. And as long as we don't check this, there is no point in checking the existence of the svg.

File: cypress/integration/diagrams.spec.ts
Comment (DerMolly) [reply cypress/integration/diagrams.spec.ts]: okay what about
describe: `rendering diagram`
	it: `markmap works`

File: cypress/integration/diagrams.spec.ts
Comment (ErikMichelson) [reply cypress/integration/diagrams.spec.ts]: Most of the libs probably have their own tests that their rendering output works. Therefore we don't need to test the libs that the use. We need to test whether we trigger the libs properly to render. When looking for a div that _could_ contain a diagram, it doesn't mean that the diagram library has been triggered. It also could be that the replacer was refactored but the lib call itself was forgotten.
When checking for the existence of just some rendering, we at least ensure that the rendering lib has been called.

File: cypress/integration/diagrams.spec.ts
Comment (ErikMichelson) [reply cypress/integration/diagrams.spec.ts]: @mrdrogdrog I don't understand why suddenly now you want to enforce your way. If we had agreed to always use such a structure or if every other test follows the same schema, it would be a different thing. But: From 131 cypress tests being in main branch less than the half - more exact 60 - are written as a sentence.

Moreover, I'll show you why I don't agree on sentences. Look at the image below. On a quick glimpse which of the tests could have failed you just see "shows correct line number" and have to read the beginning of your sentence above that.

image

File: cypress/integration/diagrams.spec.ts
Comment (mrdrogdrog) [reply cypress/integration/diagrams.spec.ts]: Theeeen how about "div > *" ? :thinking:

File: cypress/integration/diagrams.spec.ts
Comment (mrdrogdrog) [reply cypress/integration/diagrams.spec.ts]: Well. It's a pity that only jest seems to follow the convention.

We never agreed on a syntax, therefore I started to use the conventional syntax. This is why I started "now". 
Anyway, we should agree on something.

Imho the test case name should include the expectation not the object that is tested. That's what a describe block is for.

Comment (mrdrogdrog) [reply cypress/integration/diagrams.spec.ts]: like `"Diagram embeddings" + "renders markmap"`

File: cypress/integration/diagrams.spec.ts
Comment (mrdrogdrog) [reply cypress/integration/diagrams.spec.ts]: In cypress it should be
```
      .children()
      .should("exist")
```

Comment (ErikMichelson) [reply cypress/integration/diagrams.spec.ts]: I did this with the `.children()` method you mentioned.

File: cypress/integration/diagrams.spec.ts
Comment (ErikMichelson) [reply cypress/integration/diagrams.spec.ts]: > like `"Diagram embeddings" + "renders markmap"`

Did that.

**Review by mrdrogdrog — APPROVED** [review]

**Pull request merged**
